### PR TITLE
Store span in ASGI scope

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
## Motivation

In the Django integration we'd like to be able to access the ASGI
span from the scope of the request to modify it with additional tags
which aren't accessible from the span modifier.

### CODE REVIEW

File: tests/contrib/asgi/test_asgi.py [hunk 1]
```diff
@@ -362,3 +363,42 @@ async def test_bad_headers(scope, tracer, test_spans):
         "type": "http.response.body",
         "body": b"*",
     }
+
+
+async def test_get_asgi_span(tracer, test_spans):
```
Comment (brettlangdon) [inline tests/contrib/asgi/test_asgi.py:368]: can we get a test case for `span_from_scope` if the span is not in the scope?

**Review by brettlangdon — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

File: ddtrace/contrib/asgi/middleware.py [hunk 2]
```diff
@@ -104,6 +119,8 @@ async def __call__(self, scope, receive, send):
             span_type=SpanTypes.WEB,
         )
 
+        scope["datadog_span"] = span
```
Comment (majorgreys) [inline ddtrace/contrib/asgi/middleware.py:122]: Two suggestions: add a "datadog" dict we can use moving forward and specify the span as the "request" (aka root level span) span.

```suggestion
        scope["datadog"] = {"request_span": span}
```

**Review by majorgreys — COMMENTED** [review]
We discussed the potential problem of memory leaks with adding the scope this way. I wonder if we should take the safest route here and be passing a weak reference to the span. And/or we do some cleanup when `span.finish()` is called to remove it from the `scope`.

**Comment by Kyle-Verhoog** [comment]
Yeah I'm not too concerned about the possibility of memory leaks as this would apply to anything else in the ASGI scope but we can clean it up in the `finally` where the span is finished :+1:

File: ddtrace/contrib/asgi/middleware.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/asgi/middleware.py:122]: :+1:

Comment (Kyle-Verhoog) [reply ddtrace/contrib/asgi/middleware.py:122]: Done in #2726

**Comment by majorgreys** [comment]
@Mergifyio backport 0.51

**Pull request merged**